### PR TITLE
Add RPC support (services) [15364]

### PR DESCRIPTION
**Pull request by juanlofer-eprosima**
Merge after #249 

Tasks:
- [X] Implement RPC support for multiple clients
- [X] Implement RPC support for multiple servers (in touch with same or different DDS-Router participants)
- [x] Dynamically enable/disable services depending on the availability of servers
- [x] Dynamically enable/disable services according to allowlist/blocklist (after configuration reload)
- [x] Support repeater configuration option
- [ ] Add docker-compose blackbox tests (with and without repeater)
- [ ] (optional) Add new testing service example developed with Fast-DDS library
- [x] Hard-code default service QoS (temporal)
- [x] Improve error handling
- [x] Improve log traces
- [ ] Document code
- [ ] Modify documentation if required
- [x] Release notes

Caveats:
- No synchronization between matching threads of proxy servers and clients. Add if frequent race conditions arise. Note that it is only possible to reduce the probability of occurrence, but it cannot be completely avoided.

### CODE REVIEW

File: ddsrouter_core/src/cpp/communication/rpc/RPCBridge.cpp [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+        std::shared_ptr<ParticipantsDatabase> participants_database,
+        std::shared_ptr<PayloadPool> payload_pool,
+        std::shared_ptr<utils::SlotThreadPool> thread_pool,
+        bool enable /* = false */)
+    : Bridge(participants_database, payload_pool, thread_pool, enable)
+    , topic_(topic)
+{
+    logDebug(DDSROUTER_RPCBRIDGE, "Creating RPCBridge " << *this << ".");
+
+    std::set<ParticipantId> ids = participants_->get_rtps_participants_ids();
```
Comment (jparisu) [inline ddsrouter_core/src/cpp/communication/rpc/RPCBridge.cpp:45]: Why is this needed? Why non RTPS Participants could not work with services?

File: ddsrouter_core/src/cpp/communication/rpc/RPCBridge.cpp [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+            }
+
+            ret = reply_writers_[registry_entry.first]->write(data, registry_entry.second);
+
+            if (!ret)
+            {
+                // logWarning(DDSROUTER_RPCBRIDGE, "Error writting data in Track " << topic_ << ". Error code "
+                //                                                             << ret <<
+                //         ". Skipping data for this writer and continue.");
+                payload_pool_->release_payload(data->payload);
```
Comment (jparisu) [inline ddsrouter_core/src/cpp/communication/rpc/RPCBridge.cpp:313]: I would try to join every release_payload in one call (last one) and intermediate "error cases" all flow to it.
This is a common source of errors where changing/removing this `release`clause will create very hard trackable errors

File: ddsrouter_core/src/cpp/communication/rpc/RPCBridge.cpp [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+                continue;
+            }
+            service_registries_[reader->participant_id()]->erase(data->write_params.sample_identity().sequence_number());
```
Comment (jparisu) [inline ddsrouter_core/src/cpp/communication/rpc/RPCBridge.cpp:316]: `service_registries_` is not thread safe

File: ddsrouter_core/src/cpp/writer/implementations/rtps/rpc/RequestWriter.cpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    reply_related_sample_identity.sequence_number(data->sequenceNumber);
+    // if (reply_related_sample_identity == SampleIdentity::Unknown())
+    // {
+    //     logWarning();
+        // return utils::ReturnCode::RETCODE_ERROR;
+    // }
+
+    // Thread safe as registry's related_sample_identity is set only in creation, when this writer is still disabled
+    // Also note that concurrent \c write_ operations in the same \c RequestWriter are not possible, as \c write from parent \c BaseWriter is guarded
+    wparams.related_sample_identity(service_registry_->related_sample_identity_nts());
```
Comment (jparisu) [inline ddsrouter_core/src/cpp/writer/implementations/rtps/rpc/RequestWriter.cpp:92]: Could not this be done by every Writer?
Doing so and adding to service_registry when calling `write` and not here will avoid creating and maintaining this new class

File: ddsrouter_core/src/cpp/writer/implementations/rtps/rpc/RequestWriter.hpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+#include <ddsrouter_core/types/topic/RPCTopic.hpp>
+
+#include <writer/implementations/rtps/Writer.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+namespace core {
+namespace rtps {
+
+class RequestWriter : public Writer
```
Comment (jparisu) [inline ddsrouter_core/src/cpp/writer/implementations/rtps/rpc/RequestWriter.hpp:33]: For `Reply` and `RequestWirter`. This is very difficult to keep specializing (for example in case of Repeater, where specific Writer is needed). Couldn't this be done by generic intermediate functions?
e.g. in `rtps::Writer` have a virtual method called `set_wparams_` and these two classes only set implement this method, while other Writers has it empty. (just an idea, I have not though of consequences of this)

**Review by jparisu — CHANGES_REQUESTED** [review]

File: ddsrouter_core/src/cpp/core/DDSRouterImpl.hpp [hunk 6]
```diff
@@ -263,6 +283,9 @@ class DDSRouterImpl
      */
     std::map<types::RealTopic, bool> current_topics_;
 
+    // keep track of discovered servers and their status
+    std::map<types::RPCTopic, std::pair<bool, std::map<types::ParticipantId, std::set<types::GuidPrefix>>>> current_services_;
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddsrouter_core/src/cpp/core/DDSRouterImpl.hpp:287]: Moved servers database to RPCBridge.

[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddsrouter_core/src/cpp/core/DDSRouterImpl.hpp:287]: Hope it is easier to understand now.

File: ddsrouter_core/src/cpp/communication/rpc/ServiceRegistry.hpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            const types::RPCTopic& topic,
+            const types::ParticipantId& server_participant_id);
+
+    void enable();
+
+    void disable();
+
+    bool enabled();
+
+    void set_related_sample_identity_nts(const types::Guid& reply_reader_guid);
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddsrouter_core/src/cpp/communication/rpc/ServiceRegistry.hpp:53]: Sacrified method.

File: ddsrouter_core/src/cpp/communication/rpc/RPCBridge.hpp [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+protected:
+
+    const types::RPCTopic topic_;
+
+    // The id corresponds to that of the RTPS participant which discovered the server
+    std::map<types::ParticipantId, std::shared_ptr<ServiceRegistry>> service_registries_;
+
+    // Map readers' GUIDs to their associated thread pool tasks, and also keep a task emission flag.
+    std::map<types::Guid, std::pair<bool, utils::TaskId>> tasks_map_;
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddsrouter_core/src/cpp/communication/rpc/RPCBridge.hpp:106]: It should after the refactor, now the map with all keys is created before endpoints are, so there is no risk anymore of reading while a new element is being inserted. Besides that, concurrent reads to a map are safe (I believe).

File: ddsrouter_core/src/cpp/communication/rpc/RPCBridge.cpp (hunk 3 above)
Comment (juanlofer-eprosima) [reply ddsrouter_core/src/cpp/communication/rpc/RPCBridge.cpp:316]: Same as with `tasks_map_`.

File: ddsrouter_core/src/cpp/writer/implementations/rtps/rpc/RequestWriter.hpp (hunk 5 above)
Comment (juanlofer-eprosima) [reply ddsrouter_core/src/cpp/writer/implementations/rtps/rpc/RequestWriter.hpp:33]: I removed the new Writers and tried to do something as suggested, allowing to only have the write logic in one place.

**Comment by jparisu** [comment]
Closed in favor of #281
